panic-probe/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+//! Panic handler for `probe-run`.
+//!
+//! When this panic handler is used, panics will make `probe-run` print a backtrace and exit with a
+//! non-zero status code, indicating failure. This building block can be used to run on-device
+//! tests.
+//!
+//! This crate also overrides the Cortex-M HardFault handler: Any HardFault will also cause
+//! `probe-run` to exit with an error code.
+//!
+//! # Panic Messages
+//!
+//! By default, `panic-probe` *ignores* the panic message. You can enable one of the following
+//! features to print it instead:
+//!
+//! - `print-rtt`: Prints the panic message over plain RTT (via `rtt-target`). RTT must be
+//!   initialized by the app.
+//! - `print-defmt`: Prints the panic message via [defmt]'s transport (note that defmt will not be
+//!   used to efficiently format the message).
+//!
+//! [defmt]: https://github.com/knurling-rs/defmt/
+
+#![no_std]
+
+#[cfg(not(cortex_m))]
+compile_error!("`panic-probe` only supports Cortex-M targets (thumbvN-none-eabi[hf])");
+
+// Functionality `cfg`d out on platforms with OS/libstd.
+#[cfg(target_os = "none")]
+mod imp {
+    use core::panic::PanicInfo;
+    use core::sync::atomic::{AtomicBool, Ordering};
+
+    use cortex_m::asm;
+    use cortex_m_rt::{exception, ExceptionFrame};
+
+    #[cfg(feature = "print-rtt")]
+    use crate::print_rtt::print;
+
+    #[cfg(feature = "print-defmt")]
+    use crate::print_defmt::print;
+
+    #[cfg(not(any(feature = "print-rtt", feature = "print-defmt")))]
+    fn print(_: &core::panic::PanicInfo) {}
+
+    #[panic_handler]
+    fn panic(info: &PanicInfo) -> ! {
+        static PANICKED: AtomicBool = AtomicBool::new(false);
+
+        // Guard against infinite recursion, just in case.
+        if PANICKED.load(Ordering::Relaxed) {
+            loop {
+                asm::bkpt();
+            }
+        }
+
+        PANICKED.store(true, Ordering::Relaxed);
```

Comment:
I think you want a compile fence here otherwise the compiler is free to move this `store` to after `print` (which never refers to `PANICKED`) and that could result to `panic` being re-entered due to interrupts or panics from `print`.

Reply:
Also I think we may want to disable interrupts at the start of `panic` otherwise this scenario becomes possible:
- `main` panics; `PANICKED <- true`; we enter `print`
- interrupt (e.g. timer) fires and panics; that follows the `PANICKED = true` branch and exits `probe-run`

then `probe-run` will see an incomplete log frame / panic message

Reply:
If the panic handler reenters itself, the `load` would synchronize with the `store`, so this should only be a problem when an interrupt fires and also panics.

Reply:
(I've added a call to `interrupt::disable()` though)